aws/lambda/pytorch-auto-revert/pytorch_auto_revert/signal.py

```diff
@@ -0,0 +1,291 @@
+                            if e.is_failure:
+                                # We have a sandwich pattern
+                                return True
+                            elif e.is_pending:
+                                # We have a pending job, possible pattern, cannot confirm yet
+                                maybe = True
+
+                elif (
+                    len(oc_successes) == 1
+                    and len(oc_pending) > 1
+                    and oc_successes[0].started_at < oc_pending[-1].started_at
+                ):
+                    # If there is only one success and multiple pending jobs, we cannot confirm the sandwich
+                    if any(
+                        oc_successes[0].started_at
+                        < e.started_at
+                        < oc_pending[-1].started_at
+                        and (e.is_failure or e.is_pending)
+                        for e in nc.events
+                    ):
+                        maybe = True
+
+        return None if maybe else False
+
+    def has_loose_autorevert_pattern(self) -> bool:
+        """
+        Checks if there is subsequence of commits, where:
+        - there are two commits with a failure (not necessarily consecutive)
+        - there is at least one older commit with a success (not necessarily immediate predecessor)
+        :return:
+            True if such a pattern exists, False otherwise.
+        """
+        if len(self.commits) < 3:
+            return False
+
+        # Check for two newer commits with failure and one older commit with success
+        found_failures = 0
+        for c in self.commits:
+            if c.events_by_status(SignalStatus.FAILURE):
+                found_failures += 1
+            elif c.events_by_status(SignalStatus.SUCCESS):
+                if found_failures >= 2:
+                    return True
+                else:
+                    # potentially recovered Signal + one flake, not enough confidence
+                    return False
+
+        return False
+
+    def detect_autorevert_pattern(self) -> Optional[AutorevertPattern]:
+        """
+        Detect first autorevert pattern in the Signal.
+
+        Pattern: 3 consecutive commits where:
+        - 2 newer commits have failure
+        - 1 older commit doesn't have this failure
+
+        Note:
+            in real world relies on the previously checked invariants, such as:
+            no flakiness, no infra issues, etc.
+
+        Returns:
+            First detected autorevert pattern if exists, None otherwise.
+        """
+        # Commits are ordered newest -> older
+        if len(self.commits) < 3:
+            return None
+
+        for i in range(1, len(self.commits) - 1):
+            suspected_commit1 = self.commits[i]
+            newer_commit = self.commits[i - 1]
+            successful_base_commit = self.commits[i + 1]
+
+            if (
+                newer_commit.events_by_status(SignalStatus.FAILURE)
```

Comment:
for FAILURE we want ALL not ANY. For success we want any (as currently implemented)